src/main/java/com/example/demo/matcher/services/OrderService.java

```diff
@@ -1,23 +1,103 @@
 package com.example.demo.matcher.services;
 
 import com.example.demo.matcher.models.Order;
+import com.example.demo.matcher.models.OrderAction;
+import com.example.demo.matcher.models.OrderbookItem;
 
-import java.util.ArrayList;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class OrderService {
-    private ArrayList<Order> orders;
+    private final List<Order> orders;
 
-    public ArrayList<Order> get() {
+    public OrderService() {
+        orders = new ArrayList<>();
+    }
+
+    public List<Order> get() {
         return orders;
     }
 
     public void add(Order order) {
-
+        orders.add(order);
     }
 
     public void remove(Order order) {
+        orders.remove(order);
+    }
 
+    public List<OrderbookItem> getOrderbook(OrderAction action, String accountId) {
+        // filter the order list by action
+        List<Order> filtered = this.get().stream()
+                .filter(order -> order.getAction() == action && Objects.equals(order.getAccountId(), accountId))
+                .collect(Collectors.toCollection(ArrayList::new));
+        return makeOrderbook(filtered, action);
     }
 
+    public List<OrderbookItem> getOrderbook(OrderAction action) {
+        // filter the order list by action
+        ArrayList<Order> filtered = this.get().stream()
+                .filter(order -> order.getAction() == action)
+                .collect(Collectors.toCollection(ArrayList::new));
+        return makeOrderbook(filtered, action);
+    }
+
+    private static List<OrderbookItem> makeOrderbook(List<Order> orderList, OrderAction action) {
+        // aggregate orders
+        List<OrderbookItem> orderbook = aggregateOrders(orderList);
+
+        // sort increasing / decreasing by price depending on the action
+        if (action == OrderAction.BUY) orderbook.sort(Collections.reverseOrder());
+        else Collections.sort(orderbook);
+
+        return orderbook;
+    }
 
+    private static List<OrderbookItem> aggregateOrders(List<Order> orderList) {
+        // aggregate orders using hashmap
+        Map<Float, Float> aggregated = new HashMap<>();
```

Comment:
In java it's best to use bigdecimal instead of float for monetary values. This is because you lose a bit of accuracy with float.

Reply:
BigDecimals are so much worse to work with. But I see why it's important